File: DittoDatto-old/packages/shared-types/src/booking.ts
import { z } from "zod";
import { IdSchema, DateTimeSchema, PriceSchema, CurrencySchema } from "./common";

export const BookingStatusSchema = z.enum([
  "confirmed",
  "completed",
  "cancelled",
  "no-show",
  "pending",
]);
export type BookingStatus = z.infer<typeof BookingStatusSchema>;

/**
 * The Booking Schema
 */
export const BookingSchema = z.object({
  // IDENTITY & RELATIONSHIPS
  id: IdSchema,
  userId: IdSchema,
  companyId: IdSchema,
  storeId: IdSchema,
  serviceId: IdSchema,
  staffId: IdSchema.optional(), // Staff assignment
  resourceId: IdSchema.optional(), // Assigned resource (table, room, station)
  addonResourceIds: z.array(IdSchema).default([]), // Add-on resources (Beer Keg, Bartender)

  // STATUS & TIME
  status: BookingStatusSchema,
  startTime: DateTimeSchema, // "2025-12-01T14:00:00Z"
  endTime: DateTimeSchema, // "2025-12-01T14:45:00Z"
  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema.optional(),

  // FISCAL SNAPSHOT (From New Law - Critical for Receipts)
  // We save these at the moment of booking so price hikes don't affect old bookings.
  serviceTitle: z.string(),
  duration: z.number().int().positive(), // in minutes
  priceAtTimeOfBooking: PriceSchema,
  currency: CurrencySchema,

  // USER SNAPSHOT (From New Law - In case user deletes account later)
  userName: z.string(),
  userEmail: z.string().email(),
  userPhone: z.string().optional(),
  items: z.array(
    z.object({
      serviceId: IdSchema,
      title: z.string(),
      price: PriceSchema,
      duration: z.number().int().positive(),
      staffId: IdSchema.optional(),
      storeId: IdSchema,
      companyId: IdSchema,
    })
  ),

  // AI & GROUP FEATURES (From Old Law - The "Deep Dive")
  channel: z.enum(["app", "web", "voice_agent", "portal", "phone"]).default("app"), // Track booking source
  attendeeCount: z.number().int().min(1).default(1), // "Table for 4" logic
  notes: z.string().optional(),
  cancellationReason: z.string().optional(),

  // PAYMENT LINK (The Bridge to Phase 3)
  paymentId: z.string().optional(), // "pi_stripe_123" or "vipps_order_456"
});

export type Booking = z.infer<typeof BookingSchema>;

File: DittoDatto-old/services/mercury-engine-old/src/mercury_core/models/booking.py
"""Booking model — the core transactional entity.

Source: schemas/company-blueprint.surql §2.1
Zod origin: packages/shared-types/src/booking.ts

Fiscal snapshot fields (service_title, price_at_booking) are immutable at booking
time — Norwegian law requires preserving the price agreed at time of transaction.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from .common import (
    BookingChannel,
    BookingStatus,
    Currency,
    MercuryModel,
    SoftDeleteMixin,
    TimestampMixin,
)


class BookingItem(MercuryModel):
    """Individual service within a multi-service booking.

    Embeds fiscal snapshot of each service at booking time.
    """

    service: str  # record<service>
    title: str
    price: float
    duration: int  # minutes
    staff: str | None = None  # record<staff>


class Booking(MercuryModel, TimestampMixin, SoftDeleteMixin):
    """A confirmed or pending booking.

    Key design decisions:
    - Fiscal snapshot: service_title + price_at_booking frozen at booking time
    - User snapshot: user_name + user_email preserved if user deletes account
    - Multi-service: items[] array for composite bookings (same-store only)
    """

    id: str | None = None

    # Relationships (record IDs as strings)
    user_id: str = ""
    establishment: str = ""  # record<establishment>
    service: str = ""  # record<service>
    staff: str | None = None  # record<staff>
    resource: str | None = None  # record<resource>
    addon_resources: list[str] = Field(default_factory=list)

    # Status & Time
    status: BookingStatus = BookingStatus.PENDING
    start_time: datetime | None = None
    end_time: datetime | None = None

    # Fiscal snapshot (immutable — Norwegian law)
    service_title: str = ""
    duration: int = 0  # minutes
    price_at_booking: float = 0.0
    currency: Currency = Currency.NOK

    # User snapshot (preserved if user deletes account)
    user_name: str = ""
    user_email: str = ""
    user_phone: str | None = None

    # Multi-service items (embedded array)
    items: list[BookingItem] = Field(default_factory=list)

    # Metadata
    channel: BookingChannel = BookingChannel.APP
    attendee_count: int = 1
    notes: str | None = None
    cancellation_reason: str | None = None
    payment_id: str | None = None
